### Reduction operation: the sum of the numbers in the range [0, a)

In [3]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n") 

Time taken by reduction operation: 0.2850522994995117 seconds
255 ms ± 35.2 µs per loop (mean ± std. dev. of 2 runs, 1 loop each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [7]:
# a) Uso de listas
lista = list(range(value))

# a.1) Suma usando bucle for
initialTime = time.time()
suma_lista = 0
for i in lista:
    suma_lista += i
finalTime = time.time()
print("Time with list + for:", finalTime - initialTime, "seconds")

# a.2) Suma usando sum()
initialTime = time.time()
suma_builtin = sum(lista)
finalTime = time.time()
print("Time with list + sum():", finalTime - initialTime, "seconds")

# Validación
print("Resultado con sum(lista):", suma_builtin)

#Aquí uso range() para crear la lista, y comparo el tiempo con for y con sum()


Time with list + for: 0.44138145446777344 seconds
Time with list + sum(): 0.03200864791870117 seconds
Resultado con sum(lista): 499999500000


In [8]:
import numpy as np

# b) Conversión a array de numpy
array = np.array(lista)

# b.1) Suma con bucle for
initialTime = time.time()
suma_np_for = 0
for i in array:
    suma_np_for += i
finalTime = time.time()
print("Time with numpy array + for:", finalTime - initialTime, "seconds")

# b.2) Suma con np.sum()
initialTime = time.time()
suma_np = np.sum(array)
finalTime = time.time()
print("Time with numpy + np.sum():", finalTime - initialTime, "seconds")

# Validación
print("Resultado con np.sum:", suma_np)

#Esto me permite comparar el rendimiento entree listas y arrays de NumPy, destacando que np.sum() es mucho más eficiente que un bucle explícito

Time with numpy array + for: 0.8982729911804199 seconds
Time with numpy + np.sum(): 0.0036170482635498047 seconds
Resultado con np.sum: 499999500000


### Comparación de rendimientos entre métodos de reducción
- **Listas en Python:**
  - Bucle `for`: ~0.44 segundos
  - Función `sum()`: ~0.03 segundos

- **Arrays con NumPy:**
  - Bucle `for`: ~0.89 segundos
  - Función `np.sum()`: ~0.0036 segundos

### Análisis
- El uso de listas con `sum()` mejora significativamente el tiempo respecto a un bucle `for`, ya que `sum()` está implementada de forma más eficiente.
- Al convertir la lista a un array de NumPy, observamos que el bucle `for` se vuelve aún más lento que con listas, debido a que los arrays requieren un acceso más estructurado en memoria, y el `for` no aprovecha la vectorización.
- Sin embargo, `np.sum()` ofrece el mejor rendimiento con diferencia. Esto se debe a que NumPy está implementado internamente en C, y sus operaciones están optimizadas al máximo.

### Conclusión
Para operaciones de reducción, el uso de NumPy junto con sus funciones vectorizadas como `np.sum()` es la opción más eficiente. Usar `sum()` sobre listas es aceptable, pero los bucles `for` son significativamente menos eficientes en comparación.
